### ISSUE TITLE
[Nextcloud 15] Nextcloud clients cannot log in after upgrading from 15.0.0 to 15.0.2

### ISSUE THREAD

**Issue by phish108** [nextcloud/server#13591]
### Confirmed clients to fail: 

Official Desktop Client:
- Nextcloud for MacOS 2.5.1
- Nextcloud for Windows 2.5.1

Mobile Clients
- Nextcloud fo iOS 2.22..7.8  (On Grant Access reports Error: Access Forbidden, Invalid Request)
- Nextcloud Android 3.4.1 (On Grant Access reports Error: Access Forbidden, Invalid Request)

### Unaffected 
- Web-client 

### Steps to reproduce

1. Install NC 15.0.0
2. Install NC client 2.5.1 
3. authenticate and sync (works nice and smooth)
4. upgrade to NC 15.0.2 
5. log client out (or reinstall client)
6. login with client
7. authenticate via password 
8. grant access (return to landing page, return to step 6 - on mobiles, stuck at error page).
9. authenticate via token 
10. grant access (return to landing page/step 9 - on mobiles, stuck at error page).

_only on mobile clients_
11. return to old method
12. enter credentials
13. return to landing screen for new accounts.

Web Login works as usual.

### Expected behaviour
Client authenticates and synchronizes the user data

### Actual behaviour
User is stuck on the client's authentication screen.

### Server configuration

**Operating system**:

alpine linux (official docker container 15-fpm-alpine)

**Web server:**
NGINX

**Database:**
MariaDB 

**PHP version:**
7.2.14

**Nextcloud version:** (see Nextcloud admin page)

15.0.2

**Updated from an older Nextcloud/ownCloud or fresh install:**

15.0.0 and fresh install

**Where did you install Nextcloud from:**

docker hub

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.1.0
  - admin_audit: 1.5.0
  - bookmarks: 0.15.1
  - calendar: 1.6.4
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 3.0.1
  - dav: 1.8.1
  - federatedfilesharing: 1.5.0
  - files: 1.10.0
  - files_pdfviewer: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_videoplayer: 1.4.0
  - firstrunwizard: 2.4.0
  - gallery: 18.2.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - nextcloud_announcements: 1.4.0
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - provisioning_api: 1.5.0
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - updatenotification: 1.5.0
  - user_ldap: 1.5.0
  - workflowengine: 1.5.0
Disabled:
  - activity
  - encryption
  - federation
  - files_external
  - files_versions
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "version": "15.0.2.0",
        "installed": true,
        "forcessl": true,
        "loglevel": 0,
        "ldapIgnoreNamingRules": false,
        "maintenance": false,
        "theme": "",
        "maxZipInputSize": 943718400,
        "allowZipDownload": true,
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "secret": "***REMOVED SENSITIVE VALUE***",
        "share_folder": "\/Shared",
        "trashbin_retention_obligation": "auto",
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "mail_smtpmode": "smtp",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "stable",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "PLAIN",
        "mail_smtpauth": 1,
        "mail_smtpport": "587",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
No (local)

**Are you using encryption:** yes/no
No (apart from HTTPS)

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
LDAP

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
+-------------------------------+-------------------------------------------------------------------------------+
| Configuration                 | |
+-------------------------------+-------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | |
| homeFolderNamingRule          | |
| lastJpegPhotoLookup           | 0 |
| ldapAgentName                 | ***|
| ldapAgentPassword             | *** |
| ldapAttributesForGroupSearch  | |
| ldapAttributesForUserSearch   | |
| ldapBackupHost                | |
| ldapBackupPort                | |
| ldapBase                      | *** |
| ldapBaseGroups                | ***|
| ldapBaseUsers                 | *** |
| ldapCacheTTL                  | 600 |
| ldapConfigurationActive       | 1 |
| ldapDefaultPPolicyDN          | |
| ldapDynamicGroupMemberURL     | |
| ldapEmailAttribute            | mail |
| ldapExperiencedAdmin          | 0 |
| ldapExpertUUIDGroupAttr       | |
| ldapExpertUUIDUserAttr        | |
| ldapExpertUsernameAttr        | |
| ldapGidNumber                 | gidNumber |
| ldapGroupDisplayName          | cn |
| ldapGroupFilter               | |
| ldapGroupFilterGroups         | |
| ldapGroupFilterMode           | 0 |
| ldapGroupFilterObjectclass    | |
| ldapGroupMemberAssocAttr      | uniqueMember |
| ldapHost                      | ldapserver |
| ldapIgnoreNamingRules         | |
| ldapLoginFilter               | (&(|(objectclass=organizationalPerson)(objectclass=inetOrgPerson))(uid=%uid)) |
| ldapLoginFilterAttributes     | |
| ldapLoginFilterEmail          | 0 |
| ldapLoginFilterMode           | 1 |
| ldapLoginFilterUsername       | 1 |
| ldapNestedGroups              | 0 |
| ldapOverrideMainServer        | 0 |
| ldapPagingSize                | 500 |
| ldapPort                      | 389 |
| ldapQuotaAttribute            | |
| ldapQuotaDefault              | |
| ldapTLS                       | 0 |
| ldapUserAvatarRule            | default |
| ldapUserDisplayName           | cn |
| ldapUserDisplayName2          | |
| ldapUserFilter                | (|(objectclass=organizationalPerson)(objectclass=inetOrgPerson)) |
| ldapUserFilterGroups          | |
| ldapUserFilterMode            | 1 |
| ldapUserFilterObjectclass     | organizationalPerson |
| ldapUuidGroupAttribute        | auto |
| ldapUuidUserAttribute         | auto |
| turnOffCertCheck              | 0 |
| turnOnPasswordChange          | 0 |
| useMemberOfToDetectMembership | 1 |
+-------------------------------+-------------------------------------------------------------------------------+
```
</details>

### Client configuration
**Browser:**
NC Client 2.5.1 (Mozilla/5.0 (Macintosh) mirall/2.5.1final (build 20181204) (Nextcloud)

**Operating system:**
Mac OS 10.14.2

#### Web server error log
<details>
<summary>Web server error log</summary>

```
Insert your webserver log here
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

The nextcloud log keeps returning these blocks from the nextcloud client.

```
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"initializing paged search for  Filter (|(objectclass=organizationalPerson)(objectclass=inetOrgPerson)) base Array\n(\n   [0] => cn=***,dc=***,dc=***\n)\n attr Array\n(\n    [0] => \n)\n limit 500 offset 0","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***,"app":"user_ldap","method":"GET","url":"\/csrftoken","message":"Ready for a paged search","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"readAttribute: cn=***,dc=***,dc=*** found","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"initializing paged search for  Filter (&(|(objectclass=organizationalPerson)(objectclass=inetOrgPerson))(uid=***)) base Array\n(\n    [0] => dc=***,dc=***\n)\n attr Array\n(\n    [0] => entryuuid\n    [1] => nsuniqueid\n    [2] => objectguid\n    [3] => guid\n    [4] => ipauniqueid\n    [5] => dn\n    [6] => uid\n    [7] => samaccountname\n    [8] => memberof\n    [9] => mail\n    [10] => cn\n    [11] => jpegphoto\n    [12] => thumbnailphoto\n)\n limit 500 offset 0","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"Ready for a paged search","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"initializing paged search for  Filter (|(objectclass=organizationalPerson)(objectclass=inetOrgPerson)) base Array\n(\n   [0] => cn=***,dc=***,dc=***\n)\n attr Array\n(\n    [0] => \n)\n limit 500 offset 0","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"Ready for a paged search","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
{"reqId":"hVhbFaW4umrmkIKpqr1S","level":0,"time":"2019-01-14T17:38:22+00:00","remoteAddr":"10.0.16.223","user":"61b6fb06-***","app":"user_ldap","method":"GET","url":"\/csrftoken","message":"readAttribute: cn=***,dc=***,dc=*** found","userAgent":"Mozilla\/5.0 (Macintosh) mirall\/2.5.1final (build 20181204) (Nextcloud)","version":"15.0.2.0"}
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

Under settings/logging I see this line repeating over and over. 

```
Error | index | Doctrine\DBAL\Schema\SchemaException: There is no column with name 'object_id' on table 'oc_activity'.
-- | -- | --
```
</details>

**Comment by victorbw** [comment]
confirmed - also having the same issue with windows client.

as far as i remember there is already an issue somewhere mentioning this, but didnt have the time to look it up...

**Comment by victorbw** [comment]
here i have it, probably related to it: https://github.com/nextcloud/desktop/issues/1018

ps: only happens with whitespaces in usernames!

**Comment by victorbw** [comment]
@phish108 what is the username - I may bet my liver it has a whitespace in it :thinking:

**Comment by phish108** [comment]
@victorbw our usernames have no whitespaces. 

I checked the backlog: 
#13262 appears to be unrelated, because 15.0.0.10 worked without problems. 
#13092 is related to webdav, we do not use webdav and the web-client works fine.
#13431 appears similar, but in our case it is not random.

**Comment by victorbw** [comment]
@phish108 so I've probably lost a liver, the only one I had :rofl: 

well, still the nc-client has to connect to the nc-server first. and if it fails its authentication at the nc-server, how will it be possible to authenticate on ldap then?

can you please double-check if the actual username (not displayname) has no whitespaces?

**Comment by danielkesselberg** [comment]
> Error | index | Doctrine\DBAL\Schema\SchemaException: There is no column with name 'object_id' on table 'oc_activity'.

https://github.com/nextcloud/activity/issues/309#issuecomment-436929111

**Comment by phish108** [comment]
@danielkesselberg Thank you for the pointer. After fixing the database, the issue remains, only the errors disappear.

**Comment by victorbw** [comment]
> @victorbw The client connects to the server. The user authenticates successfully with LDAP. The issue is at the step of granting access to the client after login. I doubt that this is related to LDAP.

so do I - most probably its on the side of ncs authentication procedure back to the client, but I'm pretty much unsure who's the one to blame.

maybe this should be copied to https://github.com/nextcloud/desktop?

**Comment by phish108** [comment]
@victorbw I don't think that it is a client or desktop related issue for 2 reasons:

1. The all clients worked before the upgrade from 15.0.0 to 15.02 but stopped working after upgrading. 
2. **All** recent clients (incl. Mobile) stopped authenticating with 15.0.2.

The mobile clients are slightly more informative and indicate a missing or broken API on the server (Access Forbidden, Invalid Request).

I have not tested 15.0.1

**Comment by phish108** [comment]
I digged further and found nothing in the logs. The access tokens are created correctly. However it seems that the redirect URL points to the wrong location or passes the wrong data.

**Comment by rullzer** [comment]
Can somebody provide me with test credentials to a system this happens. I tried it on 4 instances I have access to and all the clients work properly.

**Comment by victorbw** [comment]
> 
> 
> Can somebody provide me with test credentials to a system this happens. I tried it on 4 instances I have access to and all the clients work properly.

PN!

**Comment by victorbw** [comment]
> @victorbw The client connects to the server. The user authenticates successfully with LDAP. The issue is at the step of granting access to the client after login. I doubt that this is related to LDAP.

never said it was ldap-related ;)

**Comment by KBlixt** [comment]
https://github.com/nextcloud/android/issues/3430

TL;DR: add this to your config.php:

`'overwriteprotocol' => 'https',`

**Comment by phish108** [comment]
@KBlixt Thank you very much. This solves the issue.

However, the [documentation](https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/reverse_proxy_configuration.html#overwrite-parameters) was unclear to me, so I missed that point. Maybe also, because it worked with the initial version. I suggest to make it clearer that these settings are non-optional in reverse proxy (and maybe FPM) settings. A bulleted list just like its presented in https://github.com/nextcloud/server/blob/master/config/config.sample.php#L456-L463, instead of a big chunk of text.

@rullzer - thank you again for your support earlier. I will revoke the credentials again.
